core/txpool/validation.go

```diff
@@ -253,39 +252,33 @@ func ValidateTransactionWithState(tx *types.Transaction, signer types.Signer, op
 		}
 	}
 	// Ensure the transactor has enough funds to cover the transaction costs
-	var (
-		balance = opts.State.GetBalance(from).ToBig()
-		cost    = tx.Cost()
-	)
-
-	sgtBalance := new(big.Int)
-	if opts.Chainconfig != nil && opts.Chainconfig.IsOptimism() && opts.Chainconfig.Optimism.UseSoulGasToken {
-		sgtBalanceSlot := core.TargetSGTBalanceSlot(from)
-		sgtBalanceValue := opts.State.GetState(types.SoulGasTokenAddr, sgtBalanceSlot)
-		sgtBalance = new(uint256.Int).SetBytes(sgtBalanceValue[:]).ToBig()
+	balance, err := core.GetEffectiveGasBalance(opts.State, opts.Chainconfig, from, tx.Value())
+	if err != nil {
+		return fmt.Errorf("%w: balance %v, tx value %v", err, balance, tx.Value())
 	}
+	cost := tx.GasCost()
 
 	if opts.L1CostFn != nil {
 		if l1Cost := opts.L1CostFn(tx.RollupCostData()); l1Cost != nil { // add rollup cost
 			cost = cost.Add(cost, l1Cost)
 		}
 	}
-	if balance.Cmp(cost) < 0 && sgtBalance.Cmp(cost) < 0 {
-		return fmt.Errorf("%w: balance %v, sgt balance:%v, tx cost %v, overshot %v", core.ErrInsufficientFunds, balance, sgtBalance, cost, new(big.Int).Sub(cost, balance))
+	if balance.Cmp(cost) < 0 {
+		return fmt.Errorf("%w: balance %v, tx cost %v, overshot %v", core.ErrInsufficientFunds, balance, cost, new(big.Int).Sub(cost, balance))
 	}
 	// Ensure the transactor has enough funds to cover for replacements or nonce
 	// expansions without overdrafts
 	spent := opts.ExistingExpenditure(from)
 	if prev := opts.ExistingCost(from, tx.Nonce()); prev != nil {
 		bump := new(big.Int).Sub(cost, prev)
 		need := new(big.Int).Add(spent, bump)
-		if balance.Cmp(need) < 0 && sgtBalance.Cmp(need) < 0 {
```

Comment:
Looks like we're only handing `tx.value` for the latest tx, but not for those already queued up; We may also need to accumulate `tx.value` to handle it completely.

Reply:
opts.ExistingExpediture() is the function to sum up all `tx.Cost()`.  We should probably separate the cost for `tx.Value()` and `tx.GasCost()`.